docs/design/constraints.md

```diff
@@ -0,0 +1,177 @@
+
+Constraint configuration will be read when `template.json` is read and extracted to template engine cache. 
+It will be available in `ITemplateInfo`, therefore for constraint evaluation `template.json` won't be read, and cached value will be used instead. 
+New property `Constraints` will be added to `ITemplateInfo` of type `IReadOnlyList<TemplateConstraintInfo>` containing template constraints information.
+
+`TemplateConstraintInfo` has the following properties:
+- `string Type`: constraint type
+- `string Args`: JSON string containing args read from `template.json`. Args to be parsed when evaluating the constraint by constraint itself.
+
+### Types uniqueness consideration
+At the moment 3rd party constraints won't be supported. 1st party developers should ensure unique naming.
+When 3rd party constraints are supported, the way to check if the constraint type is already used should be considered.
+
+
+## Constraint component
+
+Constraint implementation should implement `ITemplateConstraintFactory` interface that creates `ITemplateConstraint`:
+```csharp
+    public interface ITemplateConstraintFactory : IIdentifiedComponent
+    {
+        /// <summary>
+        /// Unique constraint type. Should match the definition in `template.json`.
+        /// </summary>
+        string Type { get; }
+
+        /// <summary>
+        /// Creates new <see cref="ITemplateConstraint"/> based on current <see cref="IEngineEnvironmentSettings"/>.
+        /// </summary>
+        public Task<ITemplateConstraint> CreateTemplateConstraintAsync(IEngineEnvironmentSettings environmentSettings);
+    }
+```
+When creating the constraint, it is assumed that constraint will be initialized and its context will be evaluated.
+Example: if constraint needs to know the project MSBuild properties it is assumed that project will be identified and MSBuild properties will be evaluated when the constraint is being created prior to first evaluation.
+Template constraints initialization will be started in parallel once `IEngineEnvironmentSettings` are created. The methods that work with constraint will ensure task completion before evaluation.
+
+`ITemplateConstraint` definition:
+```csharp
+    public interface ITemplateConstraint
+    {
+        /// <summary>
+        /// Unique constraint type. Should match the definition in `template.json`.
+        /// </summary>
+        string Type { get; }
+
+        /// <summary>
+        /// The user friendly constraint name, that can be used in UI.
+        /// </summary>
+        string DisplayName { get; }
+
+        /// <summary>
+        /// Evaluates <see cref="ITemplateInfo"/> and returns <see cref="TemplateConstraintResult"/> containing result.
+        /// </summary>
+        TemplateConstraintResult Evaluate(string args);
```

Comment:
Can we make this EvaluateAsync?

Reply:
The reason I mention this, is for our host implementation we have this pattern where we need to transition between sync/async so that we can transition to/from the UI thread to make calls to functions which need to be called on the UI thread. This being Task<TemplateConstraintResult> EvaluateAsync would simplify this pattern.

```
   private class VisualStudioSharedTemplateEngineHost : DefaultTemplateEngineHost
   {
  
            public override bool TryGetHostParamDefault(string paramName, out string? value)
            {
                bool result = false;
                string? val = null;

                if (_hostPropertyProvider is not null)
                {
                    _threadSwitcher.AsyncBlock(async () =>
                    {
                        (result, val) = await _hostPropertyProvider.TryGetHostParamDefaultAsync(paramName);
                    });
                }

                if (result)
                {
                    value = val;
                    return result;
                }

                return base.TryGetHostParamDefault(paramName, out value);
            }

   }

```

Reply:
I'm slightly hesitant to make it async due to performance.

Constraints will be evaluated for each template when retrieving the list of matched templates, therefore it should be performant. Making it async kind of implies it might be a long-running operation that is not desirable. 

Moreover, current filtering infrastructure (in `GetTemplatesAsync`) doesn't support async in filters and all the filters are running sequentially.  Evaluating async constraints sequentially may potentially cause perf penalty, not to mention drawbacks of blocking. We would need either to refrain from handling constraints using `GetTemplateAsync` (which is handly) or rework it completely to support async filters.

For `dotnet new` the plan is to evaluate all MSBuild properties when constraint is created, and use the collection of those when constraint is evaluated (in sync manner).

Reply:
This is more a recognition that in some hosts, we may have thread affinity issues and may need to switch threads between a worker and UI thread, like I mentioned in our teams chat. For everyone else benefit, My concern here is primarily around the fact that in Visual Studio we will definitely need to make sure we are on the UI thread. 

However, we are already doing this Sync -> Async transition in the host anyway for host properties, so we could use the same pattern here.